src/hssm/likelihoods/rldm_optimized_abstraction.py

```diff
@@ -0,0 +1,276 @@
+
+# Obtain the angle log-likelihood function from an ONNX model.
+angle_logp_jax_func = make_jax_matrix_logp_funcs_from_onnx(
+    model="angle.onnx",
+)
+
+
+def annotate_function(**kwargs):
+    """Attach arbitrary metadata as attributes to a function.
+
+    Parameters
+    ----------
+    **kwargs
+        Arbitrary keyword arguments to attach as attributes.
+
+    Returns
+    -------
+    Callable
+        Decorator that adds metadata attributes to the wrapped function.
+    """
+
+    def decorator(func: Callable) -> Callable:
+        @functools.wraps(func)
+        def wrapper(*args, **inner_kwargs):
+            return func(*args, **inner_kwargs)
+
+        for key, value in kwargs.items():
+            setattr(wrapper, key, value)
+        return wrapper
+
+    return decorator
+
+
+# Inner function to compute the drift rate and update q-values for each trial.
+# This function is used with `jax.lax.scan` to process each trial in the RLDM model.
+def compute_v_trial_wise(
+    q_val: jnp.ndarray, inputs: jnp.ndarray
+) -> tuple[jnp.ndarray, jnp.ndarray]:
+    """Compute the drift rate and updates the q-values for each trial.
+
+    This function is used with `jax.lax.scan` to process each trial. It takes the
+    current q-values and the RL parameters (rl_alpha, scaler), action (response),
+    and reward (feedback) for the current trial, computes the drift rate, and
+    updates the q-values. The q_values are updated in each iteration and carried
+    forward to the next one.
+
+    Parameters
+    ----------
+    q_val
+        A length-2 jnp array containing the current q-values for the two alternatives.
+        These values are updated in each iteration and carried forward to the next
+        trial.
+    inputs
+        A 2D jnp array containing the RL parameters (rl_alpha, scaler),
+        action (response), and reward (feedback) for the current trial.
+
+    Returns
+    -------
+    tuple
+        A tuple containing the updated q-values and the computed drift rate (v).
+    """
+    rl_alpha, scaler, action, reward = inputs
+    action = jnp.astype(action, jnp.int32)
+
+    # drift rate on each trial depends on difference in expected rewards for
+    # the two alternatives:
+    # drift rate = (q_up - q_low) * scaler where
+    # the scaler parameter describes the weight to put on the difference in
+    # q-values.
+    computed_v = (q_val[1] - q_val[0]) * scaler
+
+    # compute the reward prediction error
+    delta_RL = reward - q_val[action]
+
+    # update the q-values using the RL learning rule (here, simple TD rule)
+    q_val = q_val.at[action].set(q_val[action] + rl_alpha * delta_RL)
+
+    return q_val, computed_v
+
+
+# This function computes the drift rates (v) for each subject by processing
+# their trials one by one. It uses `jax.lax.scan` to efficiently iterate over
+# the trials and compute the drift rates based on the RL parameters, actions,
+# and rewards for each trial.
+def compute_v_subject_wise(
```

Comment:
would take out `_v_` here again in the name, and also instead of `subject_wise` (which insinuates that it is done per subject but FOR EACH subject) rather `_one_subject`.